Source/Core/Core/IOS/FS/HostBackend/FS.cpp

```diff
@@ -159,179 +335,307 @@ void HostFileSystem::DoState(PointerWrap& p)
 
 ResultCode HostFileSystem::Format(Uid uid)
 {
+  if (uid != 0)
+    return ResultCode::AccessDenied;
+  if (m_root_path.empty())
+    return ResultCode::AccessDenied;
   const std::string root = BuildFilename("/");
   if (!File::DeleteDirRecursively(root) || !File::CreateDir(root))
     return ResultCode::UnknownError;
+  ResetFst();
+  SaveFst();
+  // Reset and close all handles.
+  m_handles = {};
   return ResultCode::Success;
 }
 
-ResultCode HostFileSystem::CreateFile(Uid, Gid, const std::string& path, FileAttribute, Modes)
+ResultCode HostFileSystem::CreateFileOrDirectory(Uid uid, Gid gid, const std::string& path,
+                                                 FileAttribute attr, Modes modes, bool is_file)
 {
-  std::string file_name(BuildFilename(path));
-  // check if the file already exist
-  if (File::Exists(file_name))
+  if (!IsValidNonRootPath(path) ||
+      std::any_of(path.begin(), path.end(), [](char c) { return c - ' ' > 0x5e; }))
```

Comment:
This should likely have a clarifying comment.

Reply:
Ah yeah, definitely. This is what the FS sysmodule does:

```c
  if ( len > 0 )
  {
    while ( (((u8)path[i] - 0x20) & 0xFFu) <= 0x5E )
    {
      if ( ++i >= len )
        goto path_ok;
    }
    ret = FS_INVALID;
  }
```

...and now I've noticed the check is incorrect, since c is supposed to be unsigned so as to rely on modular arithmetic. Thanks for catching this! I'll replace this with std::isprint since that's what it's really checking.